backend/src/main/java/com/google/sps/datastore/MatchDatastore.java

```diff
@@ -0,0 +1,73 @@
+package com.google.sps.datastore;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.sps.data.Match;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/** Separates datastore method calls involving Match type from caller */
+public final class MatchDatastore {
+
+  // Datastore Key/Property constants
+  private static final String KIND_MATCH = "Match";
+  private static final String PROPERTY_FIRSTPARTICIPANTUSERNAME = "firstParticipantUsername";
+  private static final String PROPERTY_SECONDPARTICIPANTUSERNAME = "secondParticipantUsername";
+  private static final String PROPERTY_DURATION = "duration";
+  private static final String PROPERTY_TIMESTAMP = "timestamp";
+
+  /** Datastore */
+  private final DatastoreService datastore;
+
+  /** Constructor that takes in DatastoreService */
+  public MatchDatastore(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /** Return entity of match */
+  private Entity getEntityFromMatch(Match match) {
+    // Set properties of entity
+    Entity entity = new Entity(KIND_MATCH);
+    entity.setProperty(PROPERTY_FIRSTPARTICIPANTUSERNAME, match.getFirstParticipantUsername());
+    entity.setProperty(PROPERTY_SECONDPARTICIPANTUSERNAME, match.getSecondParticipantUsername());
+    entity.setProperty(PROPERTY_DURATION, match.getDuration());
+    entity.setProperty(PROPERTY_TIMESTAMP, match.getTimestamp());
+    return entity;
+  }
+
+  /** Put Match in datastore and return the match key id */
+  public long addMatch(Match match) {
+    // Create and insert entity into datastore
+    Entity entity = getEntityFromMatch(match);
+    datastore.put(entity);
+
+    // Return match key id
+    return entity.getKey().getId();
+  }
+
+  /** Return Match from entity, or null if entity is null */
+  private static Match getMatchFromEntity(@Nonnull Entity entity) {
+    String firstParticipantUsername =
+        (String) entity.getProperty(PROPERTY_FIRSTPARTICIPANTUSERNAME);
+    String secondParticipantUsername =
+        (String) entity.getProperty(PROPERTY_SECONDPARTICIPANTUSERNAME);
+    int duration = ((Long) entity.getProperty(PROPERTY_DURATION)).intValue();
```

Comment:
Question: Is `Long` the same as 'long`?

Reply:
Weird datastore quirk, I can't just cast duration to an int because I think ints are stored as long in datastore

Reply:
There are three different considerations here

int and long are different because long uses a different number of bytes and can supporter larger numbers

`long` and `Long` are different because long is a primitive and `Long` is an object type that "boxes" a long. Try to use `long` wherever possible unless there's a possibility to get `null`

Reply:
The cast to a Long and then intValue() part is clunky, but I get errors when I try to cast directly to an int